# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [1]:
import numpy as np
import pandas as pd
marketing_customer_analysis_df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')
df = marketing_customer_analysis_df.copy()
df

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Open Complaints Number of Policies     Policy Type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       Vehicle Class Vehicle Size Vehicle Type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  object 
 2   State                          10279 non-null  object 
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  object 
 5   Coverage                       10910 non-null  object 
 6   Education                      10910 non-null  object 
 7   Effective To Date              10910 non-null  object 
 8   EmploymentStatus               10910 non-null  object 
 9   Gender                         10910 non-null  object 
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  object 
 12  Marital Status                 10910 non-null 

In [3]:
# Formating the column names
df = df.rename(columns={col:col.replace(' ','_').lower() for col in df.columns})
df

unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       number_of_open_complaints number_of_policies     policy_type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             policy  renew_offer_type  sales_channel  total_claim_amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       vehicle_class vehicle_size vehicle_type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [4]:
# Create a new DataFrame that only includes customers who have total_claim_amount below $1,000
low_claim_yes_response = df[
    (df.total_claim_amount <=1000) & 
(df.response == 'Yes')
]
low_claim_yes_response

unnamed:_0 customer       state  customer_lifetime_value response  \
3               3  XL78013      Oregon             22332.439460      Yes   
8               8  FM55990  California              5989.773931      Yes   
15             15  CW49887  California              4626.801093      Yes   
19             19  NJ54277  California              3746.751625      Yes   
27             27  MQ68407      Oregon              4376.363592      Yes   
...           ...      ...         ...                      ...      ...   
10844       10844  FM31768     Arizona              5979.724161      Yes   
10852       10852  KZ80424  Washington              8382.478392      Yes   
10872       10872  XT67997  California              5979.724161      Yes   
10887       10887  BY78730      Oregon              8879.790017      Yes   
10897       10897  MM70762     Arizona              9075.768214      Yes   

       coverage             education effective_to_date employmentstatus  \
3      Extended               College           1/11/11         Employed   
8       Premium               College           1/19/11         Employed   
15        Basic                Master           1/16/11         Employed   
19     Extended               College           2/26/11         Employed   
27      Premium              Bachelor           2/28/11         Employed   
...         ...                   ...               ...              ...   
10844  Extended  High School or Below            2/7/11         Employed   
10852     Basic              Bachelor           1/27/11         Employed   
10872  Extended  High School or Below            2/7/11         Employed   
10887     Basic  High School or Below            2/3/11         Employed   
10897     Basic                Master           1/26/11         Employed   

      gender  ...  number_of_open_complaints number_of_policies  \
3          M  ...                        0.0                  2   
8          M  ...                        0.0                  1   
15         F  ...                        0.0                  1   
19         F  ...                        1.0                  1   
27         F  ...                        0.0                  1   
...      ...  ...                        ...                ...   
10844      F  ...                        0.0                  3   
10852      M  ...                        0.0                  2   
10872      F  ...                        0.0                  3   
10887      F  ...                        0.0                  7   
10897      M  ...                        0.0                  8   

          policy_type        policy  renew_offer_type  sales_channel  \
3      Corporate Auto  Corporate L3            Offer2         Branch   
8       Personal Auto   Personal L1            Offer2         Branch   
15       Special Auto    Special L1            Offer2         Branch   
19      Personal Auto   Personal L2            Offer2    Call Center   
27      Personal Auto   Personal L3            Offer2          Agent   
...               ...           ...               ...            ...   
10844   Personal Auto   Personal L1            Offer2          Agent   
10852   Personal Auto   Personal L2            Offer2    Call Center   
10872   Personal Auto   Personal L3            Offer2          Agent   
10887    Special Auto    Special L2            Offer1          Agent   
10897   Personal Auto   Personal L1            Offer1          Agent   

       total_claim_amount  vehicle_class vehicle_size vehicle_type  
3              484.013411  Four-Door Car      Medsize            A  
8              739.200000     Sports Car      Medsize          NaN  
15             547.200000            SUV      Medsize          NaN  
19              19.575683   Two-Door Car        Large            A  
27              60.036683  Four-Door Car      Medsize          NaN  
...                   ...            ...          ...          ...  
10844          547.200000  F

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [5]:
list(df.columns)

['unnamed:_0',
 'customer',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size',
 'vehicle_type']

In [6]:
# filter clients who responded 'Yes'
yes_responders = df[df.response == 'Yes']
yes_responders

unnamed:_0 customer       state  customer_lifetime_value response  \
3               3  XL78013      Oregon             22332.439460      Yes   
8               8  FM55990  California              5989.773931      Yes   
15             15  CW49887  California              4626.801093      Yes   
19             19  NJ54277  California              3746.751625      Yes   
27             27  MQ68407      Oregon              4376.363592      Yes   
...           ...      ...         ...                      ...      ...   
10844       10844  FM31768     Arizona              5979.724161      Yes   
10852       10852  KZ80424  Washington              8382.478392      Yes   
10872       10872  XT67997  California              5979.724161      Yes   
10887       10887  BY78730      Oregon              8879.790017      Yes   
10897       10897  MM70762     Arizona              9075.768214      Yes   

       coverage             education effective_to_date employmentstatus  \
3      Extended               College           1/11/11         Employed   
8       Premium               College           1/19/11         Employed   
15        Basic                Master           1/16/11         Employed   
19     Extended               College           2/26/11         Employed   
27      Premium              Bachelor           2/28/11         Employed   
...         ...                   ...               ...              ...   
10844  Extended  High School or Below            2/7/11         Employed   
10852     Basic              Bachelor           1/27/11         Employed   
10872  Extended  High School or Below            2/7/11         Employed   
10887     Basic  High School or Below            2/3/11         Employed   
10897     Basic                Master           1/26/11         Employed   

      gender  ...  number_of_open_complaints number_of_policies  \
3          M  ...                        0.0                  2   
8          M  ...                        0.0                  1   
15         F  ...                        0.0                  1   
19         F  ...                        1.0                  1   
27         F  ...                        0.0                  1   
...      ...  ...                        ...                ...   
10844      F  ...                        0.0                  3   
10852      M  ...                        0.0                  2   
10872      F  ...                        0.0                  3   
10887      F  ...                        0.0                  7   
10897      M  ...                        0.0                  8   

          policy_type        policy  renew_offer_type  sales_channel  \
3      Corporate Auto  Corporate L3            Offer2         Branch   
8       Personal Auto   Personal L1            Offer2         Branch   
15       Special Auto    Special L1            Offer2         Branch   
19      Personal Auto   Personal L2            Offer2    Call Center   
27      Personal Auto   Personal L3            Offer2          Agent   
...               ...           ...               ...            ...   
10844   Personal Auto   Personal L1            Offer2          Agent   
10852   Personal Auto   Personal L2            Offer2    Call Center   
10872   Personal Auto   Personal L3            Offer2          Agent   
10887    Special Auto    Special L2            Offer1          Agent   
10897   Personal Auto   Personal L1            Offer1          Agent   

       total_claim_amount  vehicle_class vehicle_size vehicle_type  
3              484.013411  Four-Door Car      Medsize            A  
8              739.200000     Sports Car      Medsize          NaN  
15             547.200000            SUV      Medsize          NaN  
19              19.575683   Two-Door Car        Large            A  
27              60.036683  Four-Door Car      Medsize          NaN  
...                   ...            ...          ...          ...  
10844          547.200000  F

In [7]:
# Calculate metrics revenue
revenue_analysis = yes_responders.pivot_table(
    index = ['policy_type','gender'],
    values = ['monthly_premium_auto', 'customer_lifetime_value'],
    aggfunc='mean'
).round(2)
revenue_analysis

customer_lifetime_value  monthly_premium_auto
policy_type    gender                                               
Corporate Auto F                       7712.63                 94.30
               M                       7944.47                 92.19
Personal Auto  F                       8339.79                 99.00
               M                       7448.38                 91.09
Special Auto   F                       7691.58                 92.31
               M                       8247.09                 86.34

In [8]:
# Compare these insights to `total_claim_amount` patterns

claims_analysis = yes_responders.pivot_table(
    index = ['policy_type','gender'],
    values = ['total_claim_amount'],
    aggfunc='mean'
).round(2)
claims_analysis

total_claim_amount
policy_type    gender                    
Corporate Auto F                   433.74
               M                   408.58
Personal Auto  F                   452.97
               M                   457.01
Special Auto   F                   453.28
               M                   429.53

####  discuss which segments appear most profitable or low-risk for the company.

Most Profitable & Low-Risk Segments : 

Personal Auto – Female

    - Highest CLV (8,339.79) and premiums (99.00)

    - Claims not significantly higher than other segments

    - Highly attractive segment

Special Auto – Male

    - High CLV with relatively controlled claims

    - Strong long-term profitability

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [9]:
# number of customers in each state
customers_by_state = (
    df.groupby('state')['customer']
    .size()
    .reset_index(name='total_customers')
)
customers_by_state

state  total_customers
0     Arizona             1937
1  California             3552
2      Nevada              993
3      Oregon             2909
4  Washington              888

In [10]:
# states where there are more than 500 customers.
states_over_500  = customers_by_state[
    customers_by_state['total_customers'] > 500
]
states_over_500 

state  total_customers
0     Arizona             1937
1  California             3552
2      Nevada              993
3      Oregon             2909
4  Washington              888

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [11]:
# Find the maximum, minimum, and median customer lifetime value by education level and gender
clv_by_education_gender = df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['min','max','median']
).round(2)

clv_by_education_gender

min       max   median
education            gender                            
Bachelor             F       1904.00  73225.96  5640.51
                     M       1898.01  67907.27  5548.03
College              F       1898.68  61850.19  5623.61
                     M       1918.12  61134.68  6005.85
Doctor               F       2395.57  44856.11  5332.46
                     M       2267.60  32677.34  5577.67
High School or Below F       2144.92  55277.45  6039.55
                     M       1940.98  83325.38  6286.73
Master               F       2417.78  51016.07  5729.86
                     M       2272.31  50568.26  5579.10

In [12]:
# Find the maximum, minimum, and median customer lifetime value by education level and gender
clv_by_education_gender = df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(
    min_clv='min',
    max_clv='max',
    median_clv='median'
).round(2)

clv_by_education_gender

min_clv   max_clv  median_clv
education            gender                               
Bachelor             F       1904.00  73225.96     5640.51
                     M       1898.01  67907.27     5548.03
College              F       1898.68  61850.19     5623.61
                     M       1918.12  61134.68     6005.85
Doctor               F       2395.57  44856.11     5332.46
                     M       2267.60  32677.34     5577.67
High School or Below F       2144.92  55277.45     6039.55
                     M       1940.98  83325.38     6286.73
Master               F       2417.78  51016.07     5729.86
                     M       2272.31  50568.26     5579.10

#### Conclusions :

Interpretation of Customer Lifetime Value by Education Level and Gender
1. General Observations
The analysis reveals meaningful differences in customer lifetime value (CLV) across education levels and gender. *
While minimum CLV values are relatively similar across most segments, median and maximum CLV values vary significantly, indicating differences in long-term customer value and revenue concentration.

2. Median CLV Insights (Typical Customer Value)
High School or Below shows the highest median CLV for both genders:
Female: 6,039.55
Male: 6,286.73 (highest overall median)
College-educated males also demonstrate a strong median CLV (6,005.85), higher than their female counterparts.
Bachelor and Master degree holders show moderate median CLV values, generally lower than those with lower formal education.
insight:
Higher education does not necessarily correspond to higher typical customer value. 
                                                 Customers with High School or Below education levels appear to be the most consistently valuable, especially male customers.

3. Maximum CLV Insights (High-Value Customers)
The highest maximum CLV appears among:
High School or Below – Male: 83,325.38
Bachelor – Female: 73,225.96
These extreme values suggest the presence of a small number of very high-value customers within these segments.
insight:
Some segments include outliers that contribute disproportionately to total revenue, particularly among lower education levels and bachelor-degree females.

4. Minimum CLV Insights (Low-Value Customers)
Minimum CLV values are relatively close across all segments (approximately 1,900–2,400).
This suggests that low-value customers exist in every education and gender group, and no segment is entirely risk-free from low engagement or low profitability.

5. Gender-Based Patterns
In several education categories (College, High School or Below, Master), male customers exhibit higher median CLV than female customers.
In contrast, Bachelor and Doctor education levels show slightly higher median CLV among female customers.
Gender effects on CLV vary by education level, indicating that gender and education interact in shaping customer value.

6. Business Conclusions and Recommendations

Most attractive segments (typical value):
High School or Below – Male
College – Male
Segments with high upside potential (outliers):
High School or Below – Male
Bachelor – Female
Strategic implications:
Retention and upselling efforts should focus on segments with high median CLV, not only on those with high maximum values.
Marketing strategies should avoid assuming that higher education automatically implies higher customer value.
Segments with high variability in CLV may benefit from personalized offers to capture high-value customers more effectively.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

In [13]:
list(df.columns)

['unnamed:_0',
 'customer',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size',
 'vehicle_type']

In [14]:
df.dtypes

unnamed:_0                         int64
customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size    

In [15]:
df['effective_to_date']

0        2/18/11
1        1/18/11
2        2/10/11
3        1/11/11
4        1/17/11
          ...   
10905    1/19/11
10906     1/6/11
10907     2/6/11
10908    2/13/11
10909     1/8/11
Name: effective_to_date, Length: 10910, dtype: object

In [16]:
# Convert effective_to_date to datetime
df['effective_to_date'] =  pd.to_datetime(df['effective_to_date'])
df['effective_to_date']

C:\Users\davyg\AppData\Local\Temp\ipykernel_16680\3246235444.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['effective_to_date'] =  pd.to_datetime(df['effective_to_date'])


0       2011-02-18
1       2011-01-18
2       2011-02-10
3       2011-01-11
4       2011-01-17
           ...    
10905   2011-01-19
10906   2011-01-06
10907   2011-02-06
10908   2011-02-13
10909   2011-01-08
Name: effective_to_date, Length: 10910, dtype: datetime64[ns]

In [17]:
# create a new column for months
df['month'] = df['effective_to_date'].dt.month
df['month']

0        2
1        1
2        2
3        1
4        1
        ..
10905    1
10906    1
10907    2
10908    2
10909    1
Name: month, Length: 10910, dtype: int32

In [18]:
#number of policies sold by state and month

policies_by_state_month = df.pivot_table(index='state', columns='month', values=['policy'], aggfunc='count')
policies_by_state_month

policy      
month           1     2
state                  
Arizona      1008   929
California   1918  1634
Nevada        551   442
Oregon       1565  1344
Washington    463   425

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

In [19]:
# Total policies per state
policies_per_state = df.groupby('state').size().sort_values(ascending=False)
policies_per_state

state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
dtype: int64

In [20]:
# Top 3 states
top_3_states = policies_per_state.head(3).index
top_3_states

Index(['California', 'Oregon', 'Arizona'], dtype='object', name='state')

In [21]:
# Policies sold by month for top 3 states
top3_state_month_policies = (
    df[df['state'].isin(top_3_states)]
    .groupby(['state', 'month'])
    .size()
    .reset_index(name='policies_sold')
)
top3_state_month_policies

state  month  policies_sold
0     Arizona      1           1008
1     Arizona      2            929
2  California      1           1918
3  California      2           1634
4      Oregon      1           1565
5      Oregon      2           1344

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [22]:
#  analyze the effect of different marketing channels on the customer response rate.

# Create a binary response variable
df['response_binary'] = df['response'].map({'Yes': 1, 'No': 0})

# Calculate response rate by sales channel
response_rate_by_channel = (
    df.groupby('sales_channel')['response_binary']
    .mean()
    .reset_index(name='response_rate')
)

response_rate_by_channel

sales_channel  response_rate
0         Agent       0.190746
1        Branch       0.113787
2   Call Center       0.109786
3           Web       0.117141